pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/BaseClusterIntegrationTest.java

```diff
@@ -78,303 +99,478 @@
+   * @param e exception.
+   */
+  private void failure(String pqlQuery, @Nullable List<String> sqlQueries, String failureMessage,
+      @Nullable Exception e) {
+    if (GATHER_FAILED_QUERIES) {
+      if (e == null) {
+        saveFailedQuery(pqlQuery, sqlQueries, failureMessage);
+      } else {
+        saveFailedQuery(pqlQuery, sqlQueries, failureMessage, e.toString());
+      }
+    } else {
+      failureMessage += "\nPQL: " + pqlQuery;
+      if (sqlQueries != null) {
+        failureMessage += "\nSQL: " + sqlQueries;
+      }
+      if (e == null) {
+        Assert.fail(failureMessage);
+      } else {
+        Assert.fail(failureMessage, e);
+      }
     }
   }
 
-  private void saveFailedQuery(String pqlQuery, List<String> sqlQueries, String... messages) {
-    failedQueryCount++;
+  private void failure(String pqlQuery, @Nullable List<String> sqlQueries, String failureMessage) {
+    failure(pqlQuery, sqlQueries, failureMessage, null);
+  }
 
-    File file = new File(getClass().getSimpleName() + "-failed.txt");
-    PrintWriter out = null;
-    try {
-      out = new PrintWriter(new FileWriter(file, true));
+  /**
+   * Helper method to save failed queries.
+   *
+   * @param pqlQuery Pinot PQL query.
+   * @param sqlQueries H2 SQL queries.
+   * @param messages failure messages.
+   */
+  private void saveFailedQuery(String pqlQuery, @Nullable List<String> sqlQueries, String... messages) {
+    _failedQueryCount++;
+
+    try (PrintWriter printWriter = new PrintWriter(new FileWriter(_failedQueriesFile, true))) {
+      printWriter.println("PQL: " + pqlQuery);
+      if (sqlQueries != null) {
+        printWriter.println("SQL: " + sqlQueries);
+      }
+      printWriter.println("Failure Messages:");
       for (String message : messages) {
-        out.print("# ");
-        out.println(message);
-        LOGGER.warn(message);
+        printWriter.println(message);
       }
-      out.print(pqlQuery);
-      out.print("\t");
-      out.println(StringUtil.join("\t", sqlQueries.toArray(new String[sqlQueries.size()])));
-      out.println();
-      out.close();
+      printWriter.println();
     } catch (IOException e) {
-      LOGGER.warn("Failed to write to failed queries file", e);
-    } finally {
-      IOUtils.closeQuietly(out);
+      LOGGER.warn("Caught exception while writing failed query to file.", e);
     }
   }
 
-  protected void runQuery(final String pqlQuery, List<String> sqlQueries) throws Exception {
+  /**
+   * Run equivalent Pinot and H2 query and compare the results.
+   * <p>LIMITATIONS:
+   * <ul>
+   *   <li>Skip comparison for selection and aggregation group-by when H2 results are too large to exhaust.</li>
+   *   <li>Do not examine the order of result records.</li>
+   * </ul>
+   *
+   * @param pqlQuery Pinot PQL query.
+   * @param sqlQueries H2 SQL queries.
+   * @throws Exception
+   */
+  protected void runQuery(String pqlQuery, @Nullable List<String> sqlQueries)
```

Comment:
We have 3 query comparators for different use case. This one is for Pinot H2 comparison. The other two, one is compare with hardcoded results, one is compare between startree and non-startree results.